recipe/2024-01-support-private-terraform-repository.md

```diff
@@ -0,0 +1,237 @@
+
+```diff
++@secure()
++param username string
++@secure()
++param token string
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
++   recipeConfig: {
++     terraform: {
++       secrets: {
++         "dev.azure.com": {
++            username : username
++            pat: token
++          }
++       }
++     }
++   }
+    recipes: {      
+      'Applications.Datastores/mongoDatabases':{
+        default: {
+          templateKind: 'terraform'
+          templatePath: 'https://dev.azure.com/test-private-repo'
+        }
+      }
+    }
+  }
+}
+```
+But the majority of users frequently update their git personal access tokens (like once a day), and which means updating the environment resource every time the token is updated.
+
+
+Option 4 : Using Applications.Core/secretStores
+
+Use secretStore to store the username and personal access token and add the secret to the new property recipeConfig in the environment.
+
+```diff
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
++   recipeConfig: {
++     terraform: {
++       secrets: {
++         "dev.azure.com": {
++            secret: secretStore.id
++          }
++       }
++     }
++   }
+    recipes: {      
+      'Applications.Datastores/mongoDatabases':{
+        default: {
+          templateKind: 'terraform'
+          templatePath: 'https://dev.azure.com/test-private-repo'
+        }
+      }
+    }
+  }
+}
+
+resource secretStore 'Applications.Core/secretStores@2023-10-01-preview' = {
+  name: 'github'
+  properties:{
+-  	app: app.id
+    type: 'generic'
+    data: {
+      'pat': {
+        value: '<personal-access-token>'
+      }
+      'username': {
+        value: '<username>'
+      }
+    }
+  }
+}
+
+```
+But todays secret store implementation is tied to application scope and in this case, secretStore needs to be created before application and environment creation. So we can only use the secretStore resource if we extend it to have global scope(extending to environment scope would create cyclic dependency between environment and secret store).
+
+Option 4 : Using kubernetes secret
+
+Using existing kubernetes secret i.e asking the users to have the git credentials already stored in the kubernetes secret on the same cluster and use it in the recipeConfig.
+
+```diff
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
++   recipeConfig: {
++     terraform: {
++       secrets: {
++         "dev.azure.com": {
```

Comment:
Are we going to do a mapping of template paths with the keys here? How will we know if the terraform is in a private repo?

Reply:
Same question applies here too: This is a map of secrets, right? I can add any number of secrets here for each registry.

Reply:
How would this design change if we wanted to support service principals in the future? Or SSH keys?